ssh/tailssh/incubator.go

```diff
@@ -0,0 +1,281 @@
+	"golang.org/x/sys/unix"
+	"tailscale.com/cmd/tailscaled/childproc"
+	"tailscale.com/types/logger"
+)
+
+func init() {
+	childproc.Add("ssh", beIncubator)
+}
+
+var ptyName = func(f *os.File) (string, error) {
+	return "", fmt.Errorf("unimplemented")
+}
+
+var startLoginSession = func(logf logger.Logf, uid uint32, localUser, remoteUser, remoteHost, tty string) (close func() error, err error) {
+	return nil, fmt.Errorf("unimplemented")
+}
+
+// newIncubatorCommand returns a new exec.Cmd configured with
+// `tailscaled be-child ssh` as the entrypoint.
+// If tailscaled is empty, the desired cmd is executed directly.
+func newIncubatorCommand(ctx context.Context, ci *sshConnInfo, lu *user.User, tailscaled, name string, args []string) *exec.Cmd {
+	if tailscaled == "" {
+		return exec.CommandContext(ctx, name, args...)
+	}
+	remoteUser := ci.uprof.LoginName
+	if len(ci.node.Tags) > 0 {
+		remoteUser = strings.Join(ci.node.Tags, ",")
+	}
+
+	incubatorArgs := []string{
+		"be-child",
+		"ssh",
+		"--uid=" + lu.Uid,
+		"--local-user=" + lu.Name,
+		"--remote-user=" + remoteUser,
+		"--remote-ip=" + ci.src.IP().String(),
+		"--cmd=" + name,
+	}
+
+	if len(args) > 0 {
+		incubatorArgs = append(incubatorArgs, fmt.Sprintf("--cmd-args=%q", strings.Join(args, " ")))
+	}
+	return exec.CommandContext(ctx, tailscaled, incubatorArgs...)
+}
+
+const debugChild = false
+
+// beIncubator is the entrypoint to the `tailscaled be-child ssh` subcommand.
+// It is responsible for informinf the system of a new login session for the user.
+// This is typically necessary for mounting home directories and decrypting file
+// systems.
+//
+// Taiscaled launches the incubator as the same user as it was launched as.
+// The incbuator then registers a new session with the OS, sets it's own UID to
+// the specified `--uid`` and then lauches the requested `--cmd`.
+func beIncubator(args []string) error {
+	var (
+		flags      = flag.NewFlagSet("", flag.ExitOnError)
+		uid        = flags.Uint64("uid", 0, "the uid of local-user")
+		localUser  = flags.String("local-user", "", "the user to run as")
+		remoteUser = flags.String("remote-user", "", "the remote user/tags")
+		remoteIP   = flags.String("remote-ip", "", "the remote Tailscale IP")
+		ttyName    = flags.String("tty-name", "", "the tty name (pts/3)")
+		hasTTY     = flags.Bool("has-tty", false, "is the output attached to a tty")
+		cmdName    = flags.String("cmd", "", "the cmd to launch")
+		cmdArgs    = flags.String("cmd-args", "", "the args for cmd")
+	)
+	if err := flags.Parse(args); err != nil {
+		return err
+	}
+	logf := logger.Discard
+	if debugChild {
+		// The only place we can log is syslog.
+		if sl, err := syslog.New(syslog.LOG_INFO|syslog.LOG_DAEMON, "tailscaled-ssh"); err == nil {
+			logf = log.New(sl, "", 0).Printf
+		}
+	}
+
+	euid := uint64(os.Geteuid())
+	if euid == 0 {
+		// Inform the system that we are about to log someone in.
+		// We can only do this if we are running as root.
```

Comment:
On Linux, perhaps. But maybe some system/environment in the future won't have this requirement.

Maybe this euid check could also be in startLoginSession.

But we can do that later if/when needed.

Reply:
Moved to maybeStartLoginSession